File: 08.seven-segment-search/php/part2.php
<?php
$freq_map = array(
    42 => 0, 17 => 1, 34 => 2, 39 => 3, 30 => 4,
    37 => 5, 41 => 6, 25 => 7, 49 => 8, 45 => 9,
);

$count = 0;
while ($l = fgets(STDIN)) {
    $split = explode(' | ', $l);
    $pattern = explode(' ', $split[0]);
    $numbers = explode(' ', $split[1]);
    $freq = array('a' => 0, 'b' => 0, 'c' => 0, 'd' => 0, 'e' => 0, 'f' => 0, 'g' => 0);
    foreach ($pattern as $pat) {
        foreach (str_split($pat) as $c) {
            $freq[$c]++;
        }
    }
    $value = 0;
    foreach ($numbers as $number) {
        $sum = 0;
        foreach (str_split(trim($number)) as $c) {
            $sum += $freq[$c];
        }
        $value = 10 * $value + $freq_map[$sum];
    }
    $count += $value;
}
?>
<?= $count ?>

File: 08.seven-segment-search/rs/part2.rs

use std::io;
use std::io::prelude::*;

fn main() {
    let stdin = io::stdin();
    let mut count = 0;
    for line in stdin.lock().lines() {
        let line = line.unwrap();
        let inputs: Vec<&str> = line.split(" | ")
            .flat_map(|l| l.split(" ").collect::<Vec<&str>>()).collect();
        let mut freq: [i32; 7] = [0; 7];
        for i in 0..10 {
            for c in inputs[i].chars() {
                freq[c as usize - 'a' as usize] += 1;
            }
        }
        let mut value = 0;
        for i in 10..14 {
            let sum = inputs[i].chars()
                .map(|c| freq[c as usize - 'a' as usize]).sum();
            value *= 10;
            value += match sum {
                42 => 0, 17 => 1, 34 => 2,
                39 => 3, 30 => 4, 37 => 5,
                41 => 6, 25 => 7, 49 => 8,
                45 => 9, _ => 0,
            };
        }
        count += value;
    }
    println!("Result: {}", count);
}

